Teach render objects to reuse engine layers

yjbanov opened the pull request:
## Description

Teach `Layer` and its implementations, `RenderObject` and its implementations, and `PaintingContext` to reuse engine layers. The idea is that a concrete `RenderObject` creates a `Layer` and holds on to it as long as it needs it (i.e. when it is composited, and the layer type does not change). In return, each `Layer` object holds on to an `EngineLayer` and reports it to the engine via `addRetained` and `oldLayer`. This allows the Web engine to reuse DOM elements across frames. Without it, each frame drops _all_ previously rendered HTML and regenerates it from scratch.

Because render objects can manage their own layers, `RenderObject.layer` may now be called any time. This renders `RenderObject.debugLayer` unnecessary, and therefore is marked as `@Deprecated`. This getter can stay in the API indefinitely for backwards-compatibility as it is not hard to maintain.

## Issues

Fixes https://github.com/flutter/flutter/issues/37256

## Tests

I added the following tests:

* New tests in `object_test.dart` that test that layers are created and updated properly by `PaintingContext`.
* New tests in `proxy_box_test.dart` and `color_filter_test.dart` that test that all render objects that use `oldLayer` create and update layers correctly.
* New tests in `repaint_boundary_test.dart` that test that the framework manages repaint boundary `OffsetLayer` correctly.
* Changed `rendering_tester.dart` to fail test by default when `FlutterError` catches errors. Previously, a test would pass in the presence of these errors.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change ([announcement](https://groups.google.com/forum/#!topic/flutter-announce/taTeI07sK0w))
- [ ] No, this is *not* a breaking change.

## What broke?

`Layer.addToScene` now returns `void`. Previously it returned `EngineLayer`. It is now the responsibility of the `Layer` implementations to create and manage engine layers. For example, a container layer that pushes opacity previously would do this:

```dart
  @override
  ui.EngineLayer addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]) {
    ui.EngineLayer newEngineLayer = builder.pushOpacity(alpha, offset: offset + layerOffset, oldLayer: _engineLayer);
    addChildrenToScene(builder);
    builder.pop();
    return newEngineLayer;
  }
```

After this change, it would change its signature to `void` and assign the layer using the `engineLayer` setter instead of returning it:

```dart
  @override
  void addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]) {
    engineLayer = builder.pushOpacity(alpha, offset: offset + layerOffset, oldLayer: _engineLayer);
    addChildrenToScene(builder);
    builder.pop();
  }
```

yjbanov committed:
Allow RenderObject.layer to by any ContainerLayer

yjbanov committed:
PaintingContext returns a layer; a render object can store its layer

yjbanov committed:
Teach various render objects to keep layers

yjbanov committed:
void addToScene

yjbanov committed:
undo pubspecs, fix color layer

yjbanov committed:
merge

yjbanov committed:
address comments, tests, clean-up

yjbanov committed:
more oldLayer changes; markNeedsAddToScene vs alwaysNeedsAddToScene; docs

yjbanov committed:
docs on isRepaintBoundary

yjbanov committed:
more tests

yjbanov committed:
add deprecation date

yjbanov committed:
prefer const

yjbanov committed:
fix import

ferhatb commented on packages/flutter/test/rendering/rendering_tester.dart:
```diff
@@ -7,33 +7,93 @@ import 'package:flutter/gestures.dart';
+  TestRenderingFlutterBinding({ this.onErrors });
+
+  final List<FlutterErrorDetails> _errors = <FlutterErrorDetails>[];
+  VoidCallback onErrors;
```
Please add doc to explain iterating removal of errors from _errors ie reason for Iterable sync*/yield

yjbanov replied:
Done.

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -478,6 +500,37 @@ class ContainerLayer extends Layer {
+    assert(() {
+      // We should remove any layers that got added to highlight the incorrect
+      // PhysicalModelLayers. If we don't, we'll end up adding duplicate layers
+      // or potentially leaving a physical model that is now correct highlighted
```
I think the original wording is correct?

@dnfield (the original author of the doc; I'm just moving this code), WDYT?

dnfield replied:
I think either way is probably a little awkward.  Maybe

`or continuing to render stale outlines`?

yjbanov replied:
Done.

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1697,13 +1722,18 @@ class LeaderLayer extends ContainerLayer {
   /// The [offset] property must be non-null before the compositing phase of the
   /// pipeline.
-  LeaderLayer({ @required this.link, this.offset = Offset.zero }) : assert(link != null);
+  LeaderLayer({ @required LayerLink link, this.offset = Offset.zero }) : assert(link != null), _link = link;
```
Dart doesn't allow calling setters in the initializer lists.

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1799,18 +1830,23 @@ class FollowerLayer extends ContainerLayer {
   /// must be non-null before the compositing phase of the pipeline.
   FollowerLayer({
-    @required this.link,
+    @required LayerLink link,
```
Same issue here. Setters not allowed in init. lists.

yjbanov committed:
more docs in rendering_tester.dart

yjbanov committed:
fix wording about stale outlines

liyuqian commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -378,28 +389,51 @@ class PaintingContext extends ClipContext {
-  ///   matches the value of [RenderObject.needsCompositing] for the caller.
+  ///   matches the value of [RenderObject.needsCompositing] for the caller. If
+  ///   false, this method returns null, indicating that a layer is no longer
+  ///   necessary. If a render object calling this method stores the `oldLayer`
```
Should a render object always set its [RenderObject.layer] field to the result of this method to no matter if this method returns null or not?

yjbanov replied:
There's a many-to-many relationship between render objects and layers. Requiring setting it to whatever a `push*` method returns implies a one-to-one relationship. Instead, setting `RenderObject.layer` is optional. You set it to get some performance benefits, but that's it. There needs to be a logical "root" layer that's stable across frames for `RenderObject.layer` to make sense.

liyuqian commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -378,28 +389,51 @@ class PaintingContext extends ClipContext {
+      if (layer == null) {
+        layer = ClipRectLayer(clipRect: offsetClipRect, clipBehavior: clipBehavior);
+      } else {
+        layer
```
Shall we also invalidate the layer's engineLayer at this point?

yjbanov replied:
I think that's layer-specific. The old `engineLayer` may still be needed to be passed as `oldLayer` during `compositeFrame`.

liyuqian replied:
I see. So `engineLayer` could either be:
1. in a state that's synchronized with the layer so we can directly use it in `addRetained`, 
2. in a stale state where one can only use it as `oldLayer`.

Is there some mechanism that we can enforce the developer to not misuse such two states of the `engineLayer`?

yjbanov replied:
Yes, I added a bunch of assertions in the `SceneBuilder` in the engine code.

Hixie replied:
LeaderLayer shouldn't even have an engine layer, should it?

Hixie replied:
@yjbanov ^

yjbanov replied:
`LeaderLayer` sometimes creates a `TransformEngineLayer`:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/layer.dart#L1755

Hixie replied:
Oh, right.

liyuqian commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1803,28 +1887,53 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
-    OffsetLayer result;
+  // Deprecated as of July 17, 2019.
+  @Deprecated('use the "layer" getter instead.')
+  ContainerLayer get debugLayer {
```
This may be a breaking change as we're changing the return type. Maybe we can just `assert(_layer is OffsetLayer)` and return an `OffsetLayer` as this is just for backward compatibility.

yjbanov replied:
Hmm. Interesting idea. Since I'm deprecating it, might as well keep the same type. I'll give it a shot.

liyuqian reviewed (commented):
Only reviewed 4 files... Will resume reviewing tomorrow.

Hixie commented:
cc @goderbauer

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1803,28 +1887,53 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
+  /// creates an [OffsetLayer] and populates this field prior to calling the
+  /// [paint] method. The [paint] method must not replace the value of this
+  /// field.
+  ContainerLayer get layer {
```
this should be protected. Nobody outside this object should be reading it as far as I can tell.

yjbanov replied:
There are 41 unprotected usages. Most are in tests (which can be fixed by making it `@protected @visibleForTesting`). However, there are 18 usages in `package:flutter_test`, `widget_inspector.dart`, and `rendering/binding.dart`, that need this to remain public.

Hixie replied:
How many of these are for debug purposes (for which we could use a `debugLayer`) vs production?

yjbanov replied:
I was able to convert all but 5 to `debugLayer`. Here's the production code that uses `layer`:

rendering/binding.dart:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/binding.dart#L249

multiple places in widget_inspector.dart (although I'm not sure if widget inspector is considered debug code):

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/widgets/widget_inspector.dart#L559

yjbanov replied:
Looking the surrounding code in `widget_inspector.dart` it's all debug code, so I was able to convert it to `debugLayer` as well. So, the only remaining production usage of `layer` is the `MouseTracker` in `rendering/binding.dart`.

Hixie replied:
Let's move that closure into a method on RenderView, and then pass that method in rather than using a closure. That'll be generally cleaner and will allow us to make this protected.

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -1803,28 +1887,53 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// For production code, consider [layer].
-  OffsetLayer get debugLayer {
-    OffsetLayer result;
+  // Deprecated as of July 17, 2019.
```
why do we want to remove this?

yjbanov replied:
Because any render object can create and keep a layer the normal `layer` getter no longer requires `isRepaintBoundary` and `!_needsPaint`. So this getter no longer adds any value to tests. They can as successfully use the `layer` getter (in fact, I moved all our existing tests to it).

Hixie replied:
It seems to me that this should still require `needsPaint`. All the more so now than before. Indeed the setter probably should have asserts too: we presumably should never change our layer except during paint.

yjbanov replied:
In the past it required `!_needsPaint`:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/object.dart#L1816

With this change it no longer requires it, because, as you say we're producing layers during paint. Are you suggesting that we should flip the requirement to require `_needsPaint`?

yjbanov replied:
Actually, we're clearing `_needsPaint` before calling `paint`, so requiring `_needsPaint` in the `layer` getter won't work:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/object.dart#L2107

Hixie replied:
I meant that `layer` should require `!needsPaint`, sorry for the confusion above.

If we make `layer` protected I guess I'm not as worried about this. Though it sounds like we still want the `debugLayer` for those places outside of RenderObject that want to read the layer during development.

Hixie commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1733,6 +1790,34 @@ class RenderPhysicalModel extends _RenderPhysicalModelBase<RRect> {
+      ..shadowColor = shadowColor;
+    return oldLayer;
+  }
+}
```
here and in other places, might it be better to just create the layer using `oldLayer ??= PhysicalModelLayer()` and then set the attributes unconditionally? Do the constructors do anything more efficient than that?

Hixie replied:
We can give them default values, or we can allow null and document they have to be non-null before adding to scene and add the relevant asserts (we have some cases like that already). I think the ergonomics will be nicer if we remove the `@required`s and just have the one way of doing this, though. Having both ways is going to result in bugs when we accidentally only fix one code path.

liyuqian commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1710,18 +1763,22 @@ class RenderPhysicalModel extends _RenderPhysicalModelBase<RRect> {
+      context.pushLayer(layer, super.paint, offset);
       assert(() {
-        physicalModel.debugCreator = debugCreator;
+        layer?.debugCreator = debugCreator;
```
Can we have a null layer here?

liyuqian commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1804,18 +1889,22 @@ class RenderPhysicalShape extends _RenderPhysicalModelBase<Path> {
+      context.pushLayer(layer, super.paint, offset);
       assert(() {
-        physicalModel.debugCreator = debugCreator;
+        layer?.debugCreator = debugCreator;
```
Same here. It doesn't seem to me that layer could be null at this point.

liyuqian reviewed (commented):
Remaining files scanned :-)

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -108,13 +130,17 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   @override
   void dropChild(AbstractNode child) {
-    markNeedsAddToScene();
+    if (!alwaysNeedsAddToScene) {
```
It would work for `dropChild`, but it could mask bugs in other code that's not supposed to call `markNeedsAddToScene`.

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -478,6 +500,36 @@ class ContainerLayer extends Layer {
+
+  /// Consider this layer as the root and build a scene (a tree of layers)
+  /// in the engine.
+  ui.Scene buildScene(ui.SceneBuilder builder) {
```
I changed the type of `RenderObject.layer` from `OffsetLayer` to `ContainerLayer`. Moving this method makes that change less breaking. There's a couple of places where we call `layer.buildScene()`.

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1748,15 +1777,16 @@ class LeaderLayer extends ContainerLayer {
     _lastOffset = offset + layerOffset;
     if (_lastOffset != Offset.zero)
-      builder.pushTransform(Matrix4.translationValues(_lastOffset.dx, _lastOffset.dy, 0.0).storage);
+      engineLayer = builder.pushTransform(Matrix4.translationValues(_lastOffset.dx, _lastOffset.dy, 0.0).storage, oldLayer: engineLayer);
```
Yes, when a leader/follower layer is moving without changing its children, we want to reuse the engine layer DOM nodes associated with `pushTransform`.

liyuqian replied:
I see. Maybe add a comment somewhere like "The engine layer's resources may be reused through oldLayer, but do not reuse the engine layer's rasterized pixels through addRetained directly because there's no guarantee that the pixels would stay the same even if the whole subtree is unchanged."

liyuqian replied:
This is specific to the Leader/FolllowerLayer. Previously, it's impossible to re-use the engineLayer of a Leader/FollowerLayer for `addRetained` because we return null.

Now we return the engineLayer for oldLayer, but we still don't want developers to use it for `addRetained`. I don't think that we need to document every `push*` method. Just documenting `Leader/FollowerLayer.addToScene` would be enough. Maybe something like "Only use the returned engineLayer for oldLayer in .... Do not use it for addRetained".

yjbanov commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -188,7 +196,6 @@ class PaintingContext extends ClipContext {
     if (child._needsPaint) {
       repaintCompositedChild(child, debugAlsoPaintedParent: true);
     } else {
-      assert(child._layer != null);
```
In both `if` and `else` cases `child._layer` must be non-null. In fact, it must be an `OffsetLayer`. So I moved this assertion down to line 209 and made it stronger.

yjbanov commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -199,8 +206,9 @@ class PaintingContext extends ClipContext {
-    child._layer.offset = offset;
+    assert(child._layer is OffsetLayer);
+    final OffsetLayer childOffsetLayer = child._layer;
+    childOffsetLayer.offset = offset;
```
`child._layer` changed its type from `OffsetLayer` to `ContainerLayer`. With this change any render object may have a layer, not just repaint boundaries.

yjbanov committed:
address some comments

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -157,7 +183,7 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   /// corresponding engine layer, null is returned.
   @protected
-  ui.EngineLayer addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]);
+  void addToScene(ui.SceneBuilder builder, [ Offset layerOffset = Offset.zero ]);
```
I reverted the change. I only changed it because it didn't feel right. It puts the responsibility of setting the `_engineLayer` field on the _caller_ of `addToScene`. This means that the caller always has to do this:

```
childLayer._engineLayer = childLayer.addToScene(builder);
```

There are two issues with this. One is that `_engineLayer` is private and so a custom layer implementation won't be able to do this correctly. The second is that the caller needs to _remember_ to do this. They also need to make sure they do this correctly. I found a bug in `_ProxyLayer` that returned another layer's engine layer, which is wrong. It only manifested after my change because my change exercises the new assertions I added in the `SceneBuilder` API.

However, there's no rush to fix this. We can do this as a follow-up PR.

liyuqian replied:
I remembered that we've had this conversation before. Maybe let's create a Github issue to track our discussions and decisions so we won't forget it again in a few months :-)

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -157,7 +183,7 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   /// Return the engine layer for retained rendering. When there's no
```
Reverted the change.

yjbanov committed:
revert addToScene change

yjbanov commented:
@liyuqian @Hixie, I responded to your comments. PTAL.

liyuqian reviewed (approved):
LGTM modulo the documentation of engineLayer returned by Leader/FollowerLayer's addToScene method.

liyuqian commented:
Ignore my Leader/FollowerLayer comment as our offline discussions found that `alwaysNeedsAddToScene` is already set and documented for those layers.

However, it seems that https://gist.github.com/liyuqian/560c3edefe2bce147147a49db0ec4e98 will throw exceptions with this patch.

yjbanov committed:
toImage marks layer as needs addToScene

yjbanov commented:
@liyuqian 

> However, it seems that https://gist.github.com/liyuqian/560c3edefe2bce147147a49db0ec4e98 will throw exceptions with this patch.

Fixed, but let's chat about it. I'm wondering if there's a less brittle solution to this.

liyuqian commented:
@yjbanov : I see. Thanks for debugging this! I booked a 30-minute meeting on Monday just to make sure that I don't forget to discuss this with you.

yjbanov committed:
auto-track needsAddToScene

yjbanov commented:
@liyuqian I implemented auto-tracking of needsAddToScene in https://github.com/flutter/flutter/pull/36402/commits/42552bd9bf3902a63a6bdc1f3de5a214c4dfdf44. PTAL.

yjbanov committed:
repurpose ContainerLayer.toImage test

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -48,11 +48,40 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   void markNeedsAddToScene() {
+    assert(
+      !alwaysNeedsAddToScene,
+      '$runtimeType called markNeedsAddToScene.\n'
```
this line itself doesn't describe a problem. Consider something like "`$runtimeType with alwaysNeedsAddToScene set called markNeedsAddToScene.`" maybe?

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -86,6 +115,38 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+      // longer requires that its [addToScene] is called (e.g. it is now safe
+      // to add this layer as retained or use it as `oldLayer`). However, the
+      // parent must construct a new engine layer to add this layer to, and so
+      // we mark it as needing its [addToScene] called next frame.
```
what if this layer's child layer called markNeedsAddToScene on us for that exact reason?

yjbanov replied:
That is expected. This is designed to handle two situations:

1. When rendering the complete layer tree as normal. In this case we call child `addToScene` methods first, then we call `set engineLayer` for the parent. The children will call `markNeedsAddToScene` on the parent to signal that they produced new engine layers and therefore the parent needs to update. However, in this case, the parent is already updating, and so after it's done, _its_ `set engineLayer` is called and it clears the `_needsAddToScene` flag.

2. When rendering an interior layer (e.g. `OffsetLayer.toImage`). In this case we call `addToScene` for one of the children but not the parent, i.e. we produce new engine layers for children but not for the parent. Here the children will mark the parent as needing `addToScene`, but the parent does not clear the flag until some future frame decides to render it, at which point the parent knows that it cannot retain its engine layer and will call `addToScene` again.

Hixie replied:
Please document all that in the source. The existing comment wasn't clear enough for me to get that from it.

It's sad that toImage results in us throwing away the "real" engine layers.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -86,6 +115,38 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// communicate to the engine that a layer is updating a previously rendered
+  /// layer. The web engine could, for example, update the properties of
+  /// previously rendered HTML DOM nodes rather than creating new nodes.
+  ui.EngineLayer get engineLayer => _engineLayer;
```
why is the getter not protected?

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -86,6 +115,38 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// resources across frames for better app performance.
+  ///
+  /// Set this field to the value returned by one of [ui.SceneBuilder]'s "push"
+  /// methods, e.g. [ui.SceneBuilder.pushOpacity].
```
move this paragraph to the docs for the setter

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -478,6 +543,36 @@ class ContainerLayer extends Layer {
+
+  /// Consider this layer as the root and build a scene (a tree of layers)
+  /// in the engine.
+  ui.Scene buildScene(ui.SceneBuilder builder) {
```
kinda seems weird to have this on a layer at all. maybe we should move it to the pipeline?

yjbanov replied:
`buildScene` is used to render normal frames (i.e. the whole layer tree), but it's also used to render interior layers as if they were standalone trees, e.g. `toImage`. The pipeline seems to be designed to be 1-to-1 with the full layer tree.

The other issue is that moving it would be a breaking change.

Hixie replied:
I forgot about toImage. Maybe add that to the docs for this method. Thanks.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -739,7 +836,7 @@ class ContainerLayer extends Layer {
         child._addToSceneWithRetainedRendering(builder);
       } else {
-        child.addToScene(builder, childOffset);
+        child.engineLayer = child.addToScene(builder, childOffset);
```
seems weird for the parent to be setting engileLayer on the child, especially when setting it to a value that the child returned. Who owns the engineLayer?

yjbanov replied:
This is not entirely new. Callers of `addToScene` are already expected to set the `_engineLayer`  field. Here's an example in today's Flutter:

https://github.com/flutter/flutter/blob/69248258aff8ee89aacb4bc1bc211db7577b46a0/packages/flutter/lib/src/rendering/layer.dart#L176

What I'm doing in this PR is teaching more places to remember the engine layers they generated. Initially I changed it so that the layers manage their own engine layers, but then reverted it (see https://github.com/flutter/flutter/pull/36402/commits/2f7aa9ea66454ea2ab0de508d9f143a7113b5f88), deciding that breaking the API may not be worth it. However, I do prefer that layers own their engine layers and their lifecycle. I'd be on board with changing it. At this point I'd prefer doing it in a separate PR though.

liyuqian replied:
https://github.com/flutter/flutter/issues/37805 is created to track the effort to make `addToScene` return void and set the `engineLayer` inside `addToScene`.

yjbanov replied:
Thanks! I'm totally willing to make that change in a follow-up PR.

Hixie replied:
It being new or not doesn't matter. :-) We want every patch to _reduce_ technical debt.

Happy for this to happen in a later issue. Please make sure to assign it to yourself and set the milestone accordingly (and schedule it accordingly) so that we know it'll happen.

(Maybe it should happen first, before this PR?)

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1687,7 +1764,7 @@ class LayerLink {
 /// This layer collapses the accumulated offset into a transform and passes
 /// [Offset.zero] to its child layers in the [addToScene]/[addChildrenToScene]
-/// methods, so that [applyTransform] will work reliably.
+/// methods, so that [applyTransform] works reliably.
```
generally speaking I think the flutter voice would use `will work` here (conceptually the method works when called, and will be called in the future, as opposed to working in general).

That said, feel free to leave this, no big deal.

yjbanov replied:
Reverted.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1993,20 +2081,21 @@ class FollowerLayer extends ContainerLayer {
       return null; // this does not have an engine layer.
```
while you're here, can you hyperlink the class names in the docs to `alwaysNeedsAddToScene` above? thanks

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -112,19 +112,27 @@ class PaintingContext extends ClipContext {
+    if (childLayer == null) {
       assert(debugAlsoPaintedParent);
-      child._layer = OffsetLayer();
+      child._layer = childLayer = OffsetLayer();
```
seems like this should go through the setter rather than the field, since we have a setter.

that said, it seems sketchy that we have two places that can affect this field. Who owns it, the render object or the painting context or someone else? we should have a clear ownership model so we know who's job it is to update it.

yjbanov replied:
Yeah, I couldn't think of a clean _and_ non-breaking solution here. We have two kinds of ownership, which is determined by `isRepaintBoundary`.

1. A repaint boundary does not own its layer. Instead, the `PaintingContext` creates an `OffsetLayer` and assigns it to the repaint boundary `RenderObject`.
1. A non-repaint boundary owns its layer (if it's composited, of course). In that case, the concrete implementation of the `RenderObject` would create the layer and assign it to the `layer` field.

To ensure correct usage, the public `layer` setter checks that the object is not a repaint boundary. Otherwise, one could create a repaint boundary and overwrite the layer assigned by `PaintingContext`. This is why I'm not using setter here.

Hixie replied:
> I couldn't think of a clean and non-breaking solution here

In such cases the "non-breaking" part is the part we give up on. (See the breaking change policy.)

This sounds like a very confusing mess as implemented. I think we should fix this before landing. There should be a single clear documented owner for the field.

Hixie replied:
(as discussed, it's fine if we just add a comment here that says we're setting the field directly because this is the one place that IS allowed to set it when it's a repaint boundary)

yjbanov replied:
Done

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -867,45 +964,17 @@ class OffsetLayer extends ContainerLayer {
     // node. Otherwise, changing an offset layer on the very high level could
     // cascade the change to too many leaves.
-    final ui.EngineLayer engineLayer = builder.pushOffset(layerOffset.dx + offset.dx, layerOffset.dy + offset.dy);
+    final ui.EngineLayer newLayer = builder.pushOffset(layerOffset.dx + offset.dx, layerOffset.dy + offset.dy, oldLayer: _engineLayer);
```
maybe call this newEngineLayer?

It would help reduce confusion between whether it's a Layer or an EngineLayer on line 977.

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -86,6 +115,38 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// previously rendered HTML DOM nodes rather than creating new nodes.
+  ui.EngineLayer get engineLayer => _engineLayer;
+  @protected
+  set engineLayer(ui.EngineLayer newLayer) {
```
newLayer => value to avoid the confusion over whether it's a Layer or an EngineLayer

yjbanov replied:
Done.

yjbanov committed:
address comments

yjbanov commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -48,11 +48,40 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  ///
+  /// If false, the parent layer may pass this layer's engine layer to
+  /// [SceneBuilder.addRetained]. If true, [SceneBuilder.addRetained] may not
+  /// be used.
```
Done.

yjbanov commented:
@Hixie, I addressed your comments in https://github.com/flutter/flutter/pull/36402/commits/875682fa6f364a312f0341028cc783b8e49cdbba, and responded to others that I couldn't address in code. PTAL.

liyuqian commented on packages/flutter/test/rendering/layers_test.dart:
```diff
@@ -44,7 +44,7 @@ void main() {
   });
 
-  test('layer subtree dirtiness is correctly computed', () {
+  test('layer _subtreeNeedsAddToScene is correctly computed', () {
```
One more thought: if we still computes `subtreeNeedsAddToScene`, then why do we have to traverse upward during `markNeedsAddToScene`? It seems that we can still just mark the current node.

yjbanov replied:
I remove _subtreeNeedsAddToScene in favor or reusing _needsAddToScene and added tests. PTAL.

yjbanov committed:
remove _subtreeNeedsAddToScene; add tests

liyuqian reviewed (approved):
LGTM

yjbanov committed:
make layer getter protected; undeprecate debugLayer

yjbanov committed:
default constructors for layers

yjbanov commented:
It is getting increasingly harder to track down outstanding comments in the github UI. So I'm copying everything unaddressed that I could find here:

- [ ] re `RenderObject.layer` getter @Hixie says:

> Let's move that closure into a method on RenderView, and then pass that method in rather than using a closure. That'll be generally cleaner and will allow us to make this protected.

- [ ] re `RenderObject.debugLayer` getter @Hixie says:

> I meant that layer should require !needsPaint, sorry for the confusion above. If we make layer protected I guess I'm not as worried about this. Though it sounds like we still want the debugLayer for those places outside of RenderObject that want to read the layer during development.

- [ ] re required parameters in layer classes @Hixie says:

> We can give them default values, or we can allow null and document they have to be non-null before adding to scene and add the relevant asserts (we have some cases like that already). I think the ergonomics will be nicer if we remove the @requireds and just have the one way of doing this, though. Having both ways is going to result in bugs when we accidentally only fix one code path.

- [ ] re rendering the scene vs rendering interior layer nodes:

@yjbanov says:

> That is expected. This is designed to handle two situations:
> 
> 1. When rendering the complete layer tree as normal. In this case we call child `addToScene` methods first, then we call `set engineLayer` for the parent. The children will call `markNeedsAddToScene` on the parent to signal that they produced new engine layers and therefore the parent needs to update. However, in this case, the parent is already updating, and so after it's done, _its_ `set engineLayer` is called and it clears the `_needsAddToScene` flag.
> 
> 2. When rendering an interior layer (e.g. `OffsetLayer.toImage`). In this case we call `addToScene` for one of the children but not the parent, i.e. we produce new engine layers for children but not for the parent. Here the children will mark the parent as needing `addToScene`, but the parent does not clear the flag until some future frame decides to render it, at which point the parent knows that it cannot retain its engine layer and will call `addToScene` again.

@Hixie says:

> Please document all that in the source. The existing comment wasn't clear enough for me to get that from it. It's sad that toImage results in us throwing away the "real" engine layers.

- [ ] re `buildScene` being a `ContainerLayer` method instead of a `PipelineOwner` method:

@yjbanov says:

> `buildScene` is used to render normal frames (i.e. the whole layer tree), but it's also used to render interior layers as if they were standalone trees, e.g. `toImage`. The pipeline seems to be designed to be 1-to-1 with the full layer tree.

@Hixie says:

> I forgot about toImage. Maybe add that to the docs for this method. Thanks.

- [ ] re `addToScene` returning `EngineLayer` rather than owning it internally:

@yjbanov says

> This is not entirely new. Callers of `addToScene` are already expected to set the `_engineLayer`  field. Here's an example in today's Flutter:
> 
> https://github.com/flutter/flutter/blob/69248258aff8ee89aacb4bc1bc211db7577b46a0/packages/flutter/lib/src/rendering/layer.dart#L176
> 
> What I'm doing in this PR is teaching more places to remember the engine layers they generated. Initially I changed it so that the layers manage their own engine layers, but then reverted it (see https://github.com/flutter/flutter/pull/36402/commits/2f7aa9ea66454ea2ab0de508d9f143a7113b5f88), deciding that breaking the API may not be worth it. However, I do prefer that layers own their engine layers and their lifecycle. I'd be on board with changing it. At this point I'd prefer doing it in a separate PR though.

@liyuqian says

> https://github.com/flutter/flutter/issues/37805 is created to track the effort to make `addToScene` return void and set the `engineLayer` inside `addToScene`.

@Hixie says

> It being new or not doesn't matter. :-) We want every patch to _reduce_ technical debt.
> 
> Happy for this to happen in a later issue. Please make sure to assign it to yourself and set the milestone accordingly (and schedule it accordingly) so that we know it'll happen.
> 
> (Maybe it should happen first, before this PR?)

- [ ] re repaint boundaries not managing their own layers:

@yjbanov says:

> I couldn't think of a clean _and_ non-breaking solution here. We have two kinds of ownership, which is determined by `isRepaintBoundary`.
> 
> 1. A repaint boundary does not own its layer. Instead, the `PaintingContext` creates an `OffsetLayer` and assigns it to the repaint boundary `RenderObject`.
> 1. A non-repaint boundary owns its layer (if it's composited, of course). In that case, the concrete implementation of the `RenderObject` would create the layer and assign it to the `layer` field.
> 
> To ensure correct usage, the public `layer` setter checks that the object is not a repaint boundary. Otherwise, one could create a repaint boundary and overwrite the layer assigned by `PaintingContext`. This is why I'm not using setter here.

@Hixie says:

> In such cases the "non-breaking" part is the part we give up on. (See the breaking change policy.)
> 
> This sounds like a very confusing mess as implemented. I think we should fix this before landing. There should be a single clear documented owner for the field.

yjbanov committed:
add docs expalining how interior layers are rendered

yjbanov committed:
document why buildScene is on ContainerLayer

yjbanov commented:
Update: I addressed comments in the following commits

- https://github.com/flutter/flutter/pull/36402/commits/0a2455d8f3d42ac1e23c35c2e9d1d760a6d338f3 - make the `layer` getter protected; undeprecate debugLayer
- https://github.com/flutter/flutter/pull/36402/commits/98f2e5eb0585a2d1d7ceaf84dffe380a6e6e35d4 - default constructors for layers; this is so usage sites can do, e.g. `layer ??= OffsetLayer()`.
- https://github.com/flutter/flutter/pull/36402/commits/c6369e5ce5da40bee44ae89b2492dacae3ef115a, https://github.com/flutter/flutter/pull/36402/commits/00cb81538cbf450ab5f3f38ffe43944423a97ede - these two add more docs

AFAICT the remaining two issues are `Layer`-`EngineLayer` ownership and `RenderObject`-`Layer` ownership. I will attempt to resolve the latter next and update this thread.

yjbanov committed:
document why not using `layer` setter

yjbanov committed:
make addToScene void again

yjbanov commented:
I addressed the remaining two issues:

- document why `PaintingContext` is not using the `RenderObject.layer` setter (https://github.com/flutter/flutter/pull/36402/commits/bbceef190cc47fe81e395fe8b33587fd371d56cd)
- change `addToScene` to return `void` (https://github.com/flutter/flutter/pull/36402/commits/84ba5542499152e72dc6477f00bc2f71be932e1f, [breaking change announcement](https://groups.google.com/forum/#!topic/flutter-announce/taTeI07sK0w))

PTAL

Hixie commented on packages/flutter/lib/src/rendering/view.dart:
```diff
@@ -173,6 +174,19 @@ class RenderView extends RenderObject with RenderObjectWithChildMixin<RenderBox>
+    // Layer hit testing is done using device pixels, so we have to convert
+    // the logical coordinates of the event location back to device pixels
+    // here.
+    return layer.findAll<MouseTrackerAnnotation>(position * ui.window.devicePixelRatio);
```
never access `ui.window` directly; go through the binding.

Hixie replied:
actually even better, use `configuration.devicePixelRatio`

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -46,18 +46,32 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   //
   // Initialized to true as a new layer has never called [addToScene].
+  //
+  // The final value is only valid after calling [updateSubtreeNeedsAddToScene].
```
not sure what "the final value" means here

yjbanov replied:
Done. Expanded the docs.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -478,6 +548,44 @@ class ContainerLayer extends Layer {
+  /// Consider this layer as the root and build a scene (a tree of layers)
+  /// in the engine.
+  // The reason this method is in the `ContainerLayer` class rather than
+  // `PipelineOwner` or other singleton level is because this method can beused
```
beused -> be used

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -478,6 +548,44 @@ class ContainerLayer extends Layer {
+  // The reason this method is in the `ContainerLayer` class rather than
+  // `PipelineOwner` or other singleton level is because this method can beused
+  // both to render the whole layer tree (e.g. a normal application frame) as
+  // as to render a subtree (e.g. `OffsetLayer.toImage`).
```
as as -> and

yjbanov replied:
Done.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577) modulo nits

yjbanov committed:
better docs
